xcmds/cpu/cpu.go

```diff
@@ -371,19 +426,178 @@ func init() {
+		if _, err := os.Stat(v[0]); os.IsNotExist(err) {
+			verbose("it's not there")
+			continue
+		}
+
+		// I *love* special cases. Evaluate just the top-most symlink.
+		//
+		// In source mode, this would be a symlink like
+		// /buildbin/defaultsh -> /buildbin/elvish ->
+		// /buildbin/installcommand.
+		//
+		// To actually get the command to build, argv[0] has to end
+		// with /elvish, so we resolve one level of symlink.
+		if filepath.Base(v[0]) == "defaultsh" {
+			s, err := os.Readlink(v[0])
+			if err == nil {
+				v[0] = s
+			}
+			verbose("readlink of %v returns %v", v[0], s)
+			// and, well, it might be a relative link.
+			// We must go deeper.
+			d, b := filepath.Split(v[0])
+			d = filepath.Base(d)
+			v[0] = filepath.Join("/", os.Getenv("UROOT_ROOT"), d, b)
+			verbose("is now %v", v[0])
+		}
+
+		cmd := exec.Command(v[0], v[1:]...)
+		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+		cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true}
+		verbose("Run %v", cmd)
+		if err := cmd.Start(); err != nil {
+			log.Printf("Error starting %v: %v", v, err)
+			continue
+		}
+	}
+	publicKeyOption := func(ctx ssh.Context, key ssh.PublicKey) bool {
+		// Glob the users's home directory for all the
+		// possible keys?
+		data, err := ioutil.ReadFile(*pubKeyFile)
+		if err != nil {
+			fmt.Print(err)
+			return false
+		}
+		allowed, _, _, _, _ := ssh.ParseAuthorizedKey(data)
+		return ssh.KeysEqual(key, allowed)
+	}
+
+	// Now we run as an ssh server, and each time we get a connection,
+	// we run that command after setting things up for it.
+	server := ssh.Server{
+		LocalPortForwardingCallback: ssh.LocalPortForwardingCallback(func(ctx ssh.Context, dhost string, dport uint32) bool {
+			log.Println("Accepted forward", dhost, dport)
+			return true
+		}),
+		Addr:             ":" + *port,
+		PublicKeyHandler: publicKeyOption,
+		ReversePortForwardingCallback: ssh.ReversePortForwardingCallback(func(ctx ssh.Context, host string, port uint32) bool {
+			log.Println("attempt to bind", host, port, "granted")
+			return true
+		}),
+		Handler: handler,
+	}
+
+	server.SetOption(ssh.HostKeyFile(*hostKeyFile))
+	log.Println("starting ssh server on port " + *port)
+	if err := server.ListenAndServe(); err != nil {
+		log.Print(err)
+	}
+	verbose("server.ListenAndServer returned")
+
+	cpuDone()
+	return nil
+}
+
 func main() {
+	verbose("Args %v pid %d *runasinit %v *remote %v", os.Args, os.Getpid(), *runAsInit, *remote)
 	a := strings.Join(flag.Args(), " ")
-	if *remote {
+	switch {
+	case *runAsInit:
+		verbose("Running as Init")
+		if err := doInit(); err != nil {
+			log.Fatal(err)
+		}
+	case *remote:
+		verbose("Running as remote")
 		if err := runRemote(a, *port9p); err != nil {
 			log.Fatal(err)
 		}
-		return
-	}
-	if a == "" {
-		a = os.Getenv("SHELL")
-	}
-	if err := runClient(a); err != nil {
-		log.Fatal(err)
+	default:
+		verbose("Running as client")
+		if a == "" {
+			a = os.Getenv("SHELL")
+		}
```

Comment:
Should you check that a isn't empty?

Reply:
I decided not to. I've found no answer that's universally correct and the convention is that if it is not set, it's up to the program.